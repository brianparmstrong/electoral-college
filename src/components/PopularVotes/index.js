import { arrayOf, bool, func, number, oneOfType, shape, string } from 'prop-types';
import PopularVoteInput from '../PopularVoteInput';
import { PARTY_MAP } from '../../constants';

const PopularVotes = ({
  currentPVTotals,
  evs,
  handlePropVotes,
  name,
  popVotesData,
  showPopVotes,
  stateEvs
}) => {
  const popVotesDataValues = popVotesData ? popVotesData.values : [0, 0, 0, 0, 0];
  const wrapperClass = showPopVotes ? '' : 'hide';

  const renderPopVoteInput = () => {
    return popVotesDataValues.map((value, i) => {
      return (
        <PopularVoteInput
          currentPVTotals={currentPVTotals}
          evs={evs}
          handlePropVotes={handlePropVotes}
          key={`${name}-${PARTY_MAP[i]}`}
          name={name}
          party={PARTY_MAP[i]}
          percent={value}
          stateEvs={stateEvs} />
      );
    });
  };

  return (
    <div className={`popVotesWrapper ${wrapperClass}`}>
      {renderPopVoteInput()}
    </div>
  ); 
};

PopularVotes.propTypes = {
  currentPVTotals: arrayOf(number).isRequired,
  evs: oneOfType([number, string]).isRequired,
  handlePropVotes: func.isRequired,
  name: string.isRequired,
  popVotesData: shape({
    name: string,
    values: arrayOf(string).isRequired
  }),
  showPopVotes: bool.isRequired,
  stateEvs: number
};

export default PopularVotes;
