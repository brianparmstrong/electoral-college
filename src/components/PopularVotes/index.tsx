import { useEffect, useState } from 'react';
import PopularVoteInput from '../PopularVoteInput';
import { PARTY_MAP } from '../../constants';
import { IfcPopularVotes } from '../../types';

const PopularVotes = (popularVotes: IfcPopularVotes) => {
  const {
    currentPVTotals,
    evs,
    handlePropVotes,
    hasClearedSavedData,
    name,
    popVotesData,
    showPopVotes,
    stateEvs,
  } = popularVotes;
  const popVotesDataValues = popVotesData
    ? popVotesData.values
    : ['0', '0', '0', '0', '0'];
  const wrapperClass = showPopVotes ? '' : 'hide';
  const [clearInputs, setClearInputs] = useState<'false' | 'true'>(
    hasClearedSavedData
  );

  useEffect(() => {
    setClearInputs(hasClearedSavedData);
  }, [hasClearedSavedData]);

  const renderPopVoteInput = () => {
    return popVotesDataValues.map((value: string, i: number) => {
      return (
        <PopularVoteInput
          currentPVTotals={currentPVTotals}
          evs={evs}
          handlePropVotes={handlePropVotes}
          hasClearedSavedData={clearInputs}
          key={`${name}-${PARTY_MAP[i]}`}
          name={name}
          party={PARTY_MAP[i]}
          percent={clearInputs === 'true' ? '0' : popVotesData && value}
          stateEvs={stateEvs}
        />
      );
    });
  };

  return (
    <div className={`popVotesWrapper ${wrapperClass}`}>
      {renderPopVoteInput()}
    </div>
  );
};

export default PopularVotes;
