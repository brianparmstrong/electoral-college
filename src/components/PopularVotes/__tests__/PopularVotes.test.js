import React from 'react';
import renderer from 'react-test-renderer';
import PopularVotes from '../';

const MOCK_POP_VOTES_PROPS = {
  currentPVTotals: [ 100, 90, 5, 2, 1 ],
  evs: 1,
  handlePropVotes: jest.fn(),
  name: 'ME-CD1',
  popVotesData: { values: [ '', '', '', '', '' ] },
  showPopVotes: true,
  stateEvs: 4
};

it('renders correctly', () => {
  const popVotes = renderer
    .create(
      <PopularVotes { ...MOCK_POP_VOTES_PROPS } />
    )
    .toJSON();
  expect(popVotes).toMatchSnapshot();
});
