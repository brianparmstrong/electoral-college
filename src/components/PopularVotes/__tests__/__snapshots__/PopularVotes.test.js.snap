// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="popVotesWrapper "
>
  <div
    className="pvInputWrapper"
  >
    <input
      className="pvInput"
      data-evsawarded="undefined"
      data-party="gop"
      data-statename="ME-CD1"
      defaultValue=""
      id="gopPopVoteInput"
      onBlur={[Function]}
      onFocus={[Function]}
      pattern="^(100(?:\\\\.00)?|0(?:\\\\.\\\\d\\\\d)?|\\\\d?\\\\d(?:\\\\.\\\\d\\\\d)?)$"
      placeholder="GOP %"
      type="text"
    />
  </div>
  <div
    className="pvInputWrapper"
  >
    <input
      className="pvInput"
      data-evsawarded="undefined"
      data-party="dem"
      data-statename="ME-CD1"
      defaultValue=""
      id="demPopVoteInput"
      onBlur={[Function]}
      onFocus={[Function]}
      pattern="^(100(?:\\\\.00)?|0(?:\\\\.\\\\d\\\\d)?|\\\\d?\\\\d(?:\\\\.\\\\d\\\\d)?)$"
      placeholder="DEM %"
      type="text"
    />
  </div>
  <div
    className="pvInputWrapper"
  >
    <input
      className="pvInput"
      data-evsawarded="undefined"
      data-party="lib"
      data-statename="ME-CD1"
      defaultValue=""
      id="libPopVoteInput"
      onBlur={[Function]}
      onFocus={[Function]}
      pattern="^(100(?:\\\\.00)?|0(?:\\\\.\\\\d\\\\d)?|\\\\d?\\\\d(?:\\\\.\\\\d\\\\d)?)$"
      placeholder="LIB %"
      type="text"
    />
  </div>
  <div
    className="pvInputWrapper"
  >
    <input
      className="pvInput"
      data-evsawarded="undefined"
      data-party="grn"
      data-statename="ME-CD1"
      defaultValue=""
      id="grnPopVoteInput"
      onBlur={[Function]}
      onFocus={[Function]}
      pattern="^(100(?:\\\\.00)?|0(?:\\\\.\\\\d\\\\d)?|\\\\d?\\\\d(?:\\\\.\\\\d\\\\d)?)$"
      placeholder="GRN %"
      type="text"
    />
  </div>
  <div
    className="pvInputWrapper"
  >
    <input
      className="pvInput"
      data-evsawarded="undefined"
      data-party="ind"
      data-statename="ME-CD1"
      defaultValue=""
      id="indPopVoteInput"
      onBlur={[Function]}
      onFocus={[Function]}
      pattern="^(100(?:\\\\.00)?|0(?:\\\\.\\\\d\\\\d)?|\\\\d?\\\\d(?:\\\\.\\\\d\\\\d)?)$"
      placeholder="IND %"
      type="text"
    />
  </div>
</div>
`;
